File: modules/kscan-live-vto-native/ios/Drivers/LiveVtoReplayDriver.swift
import Foundation
import os.log

/// Owns the replay CLOCK -- the one thing `LiveVtoReplaySession` deliberately
/// does not own, so the session stays a pure, host-testable state machine.
/// Field-for-field port of Android's `LiveVtoReplayDriver.kt`.
///
/// ── Thread topology ──────────────────────────────────────────────────────
///
///   kscan-live-vto-replay   (this class, one serial background queue)
///       frame acquisition from the replay source
///       BodyFrame production
///       deformation / geometry compute (LiveVtoGeometryPipeline)
///       publish into LatestStateSlot
///
///   UIKit draw thread
///       LiveVtoRenderView.draw(_:) -> slot.peek() -> Core Graphics
///
///   Bridge / event dispatch
///       ReplayEvent callbacks, on the replay queue; the module marshals
///       them onward. State transitions only -- never per frame.
///
/// NOTHING in the first group runs on the main thread. Rasterization runs on
/// the view's own draw call, which is correct and required for UIKit's
/// rendering model -- what must stay off it is production and deformation,
/// and both live here.
///
/// This class deliberately does NOT use a repeating `DispatchSourceTimer`
/// (which schedules at a FIXED RATE and can fire a burst of catch-up ticks
/// after a slow one -- exactly the queueing behavior a fixed-DELAY clock
/// forbids). Instead each tick reschedules itself `framePeriod` after it
/// FINISHES, matching Android's deliberate choice of
/// `scheduleWithFixedDelay` over `scheduleAtFixedRate`.
public final class LiveVtoReplayDriver {
  public static let threadName = "kscan-live-vto-replay"

  /// ~30 production ticks per second. Deliberately NOT tuned to any device
  /// or simulator: this exists to prove the architecture drops rather than
  /// queues; real cadence is re-measured against real perception and camera
  /// workloads once camera work begins.
  public static let defaultFramePeriodSeconds: TimeInterval = 0.033

  private let session: LiveVtoReplaySession
  private let framePeriodSeconds: TimeInterval
  private let queue = DispatchQueue(label: LiveVtoReplayDriver.threadName, qos: .userInitiated)
  private let runningLock = NSLock()
  private var running = false
  /// Bumped on every `stop()`; a scheduled tick checks its own generation
  /// before running and reschedules only if unchanged, so a `stop()` racing
  /// an in-flight `asyncAfter` cannot resurrect the clock.
  private var generation: Int = 0

  public init(session: LiveVtoReplaySession, framePeriodSeconds: TimeInterval = LiveVtoReplayDriver.defaultFramePeriodSeconds) {
    self.session = session
    self.framePeriodSeconds = framePeriodSeconds
  }

  /// Starts the clock. Idempotent -- a second call while running is a no-op.
  public func start() {
    runningLock.lock()
    guard !running else { runningLock.unlock(); return }
    running = true
    let myGeneration = generation
    runningLock.unlock()
    queue.async { [weak self] in self?.tick(generation: myGeneration) }
  }

  private func tick(generation myGeneration: Int) {
    runningLock.lock()
    guard running, generation == myGeneration else { runningLock.unlock(); return }
    runningLock.unlock()

    _ = session.advance()

    runningLock.lock()
    guard running, generation == myGeneration else { runningLock.unlock(); return }
    runningLock.unlock()
    queue.asyncAfter(deadline: .now() + framePeriodSeconds) { [weak self] in self?.tick(generation: myGeneration) }
  }

  /// Stops the clock. Returns once no further tick will run for the current generation.
  public func stop() {
    runningLock.lock()
    guard running else { runningLock.unlock(); return }
    running = false
    generation += 1
    runningLock.unlock()
  }

  public var isRunning: Bool { runningLock.lock(); defer { runningLock.unlock() }; return running }
}

File: modules/kscan-live-vto-native/android/src/main/java/expo/modules/kscanlivevtonative/LiveVtoReplayDriver.kt
package expo.modules.kscanlivevtonative

import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Owns the replay CLOCK -- the one thing `LiveVtoReplaySession` deliberately
 * does not own, so the session stays a pure, JVM-testable state machine.
 *
 * ── Thread topology (amendment D18) ───────────────────────────────────────
 *
 *   kscan-live-vto-replay   (this class, one daemon thread)
 *       frame acquisition from the replay source
 *       BodyFrame production
 *       deformation / geometry compute (LiveVtoGeometryPipeline)
 *       publish into LatestStateSlot
 *
 *   Android UI / View draw thread
 *       LiveVtoTestRenderView.onDraw -> slot.peek() -> Canvas.drawBitmapMesh
 *
 *   Bridge / event dispatch
 *       ReplayEvent callbacks, on the replay thread; the module marshals
 *       them onward. State transitions only -- never per frame.
 *
 * NOTHING in the first group runs on the UI thread. Canvas rasterization
 * runs on the View's own draw thread, which is correct and required for the
 * Android View rendering model (amendment D10) -- what must stay off it is
 * production and deformation, and both live here.
 *
 * The scheduler uses `scheduleWithFixedDelay`, not `scheduleAtFixedRate`:
 * a slow tick must not cause the executor to fire a burst of catch-up ticks,
 * which would be a queue by another name.
 */
class LiveVtoReplayDriver(
  private val session: LiveVtoReplaySession,
  private val framePeriodMillis: Long = DEFAULT_FRAME_PERIOD_MS,
) {
  private val running = AtomicBoolean(false)
  private var executor: ScheduledExecutorService? = null
  private var task: ScheduledFuture<*>? = null

  /** Starts the clock. Idempotent -- a second call while running is a no-op. */
  fun start() {
    if (!running.compareAndSet(false, true)) return
    val exec = Executors.newSingleThreadScheduledExecutor { runnable ->
      Thread(runnable, THREAD_NAME).apply { isDaemon = true }
    }
    executor = exec
    task = exec.scheduleWithFixedDelay({
      try {
        session.advance()
      } catch (t: Throwable) {
        // A production tick must never kill the executor thread: a
        // ScheduledExecutorService silently cancels a repeating task whose
        // body throws, which would look exactly like a clean stop while the
        // session still reported PLAYING.
        android.util.Log.e(THREAD_NAME, "replay tick failed", t)
      }
    }, 0L, framePeriodMillis, TimeUnit.MILLISECONDS)
  }

  /**
   * Stops the clock and waits for the in-flight tick to finish, so no
   * production can land after a caller believes the driver has stopped.
   */
  fun stop() {
    if (!running.compareAndSet(true, false)) return
    task?.cancel(false)
    task = null
    val exec = executor ?: return
    executor = null
    exec.shutdown()
    if (!exec.awaitTermination(2, TimeUnit.SECONDS)) exec.shutdownNow()
  }

  val isRunning: Boolean get() = running.get()

  companion object {
    const val THREAD_NAME = "kscan-live-vto-replay"

    /**
     * ~30 production ticks per second. Deliberately NOT tuned to any device
     * or emulator: N1-D exists to prove the architecture drops rather than
     * queues, and real cadence is re-measured against real perception and
     * camera workloads at N1-E/F (amendment D15).
     */
    const val DEFAULT_FRAME_PERIOD_MS = 33L
  }
}
